engine/schema/src/main/java/com/cloud/vm/dao/VMInstanceDaoImpl.java

```diff
@@ -802,6 +802,27 @@ public Long countByZoneAndState(long zoneId, State state) {
         return customSearch(sc, null).get(0);
     }
 
+    @Override
+    public Long countByZoneAndStateAndHostTag(long dcId, State state, String hostTag) {
+        TransactionLegacy txn = TransactionLegacy.currentTxn();
+        String sql = "SELECT COUNT(1) FROM vm_instance vi "
+                + "JOIN service_offering so ON vi.service_offering_id=so.id "
+                + "JOIN vm_template vt ON vi.vm_template_id = vt.id "
+                + "WHERE vi.data_center_id= " + dcId + " AND vi.state = '" + state + "' AND vi.removed IS NULL "
+                + "AND (so.host_tag='" + hostTag + "' OR vt.template_tag='" + hostTag + "')";
```

Comment:
Why not define this with prepared statement syntax? these parameters are very suitable for that
```suggestion
        String sql = "SELECT COUNT(1) FROM vm_instance vi "
                + "JOIN service_offering so ON vi.service_offering_id=so.id "
                + "JOIN vm_template vt ON vi.vm_template_id = vt.id "
                + "WHERE vi.data_center_id= ? AND vi.state = ? AND vi.removed IS NULL "
                + "AND (so.host_tag = ? OR vt.template_tag = ? )";
```
and later assign `dcId`, `state`, `hostTag` and again `state`. It is saver and if executed a second time in the same db engine also quicker.

Reply:
@soreana can you answer me on my question above?

Reply:
I'm working on this one. I'm not good at writing search criteria. I hope I finish that later today or tomorrow. 🤞

Reply:
@DaanHoogland I gave up using search criteria and used prepared statements. SearchCriteria with two join statements were over complicated for me.